clippy_lints/src/matches/needless_match.rs

```diff
@@ -116,44 +126,68 @@ fn strip_return<'hir>(expr: &'hir Expr<'hir>) -> &'hir Expr<'hir> {
     }
 }
 
+/// Check if the type of the match operand differs with the assigned local or function return type
+///
+/// Can also be used to check for the `let_expr` in `IfLet` as well.
+fn is_type_differ(cx: &LateContext<'_>, match_expr: &Expr<'_>, expr: &Expr<'_>) -> bool {
+    if let Some(p_node) = get_parent_node(cx.tcx, expr.hir_id) {
```

Comment:
We have `clippy_utils::is_adjusted(cx, expr)`, which should do the trick more easily.

Reply:
> We have `clippy_utils::is_adjusted(cx, expr)`, which should do the trick more easily.

Well... I later realized that I might still need to keep it this way, because I do need to make sure that the type of match input and output are the same, not just coercion but also as what #8599 suggested. unless there is another util function that I can use to do the check.